qcodes/tests/dataset/test_data_set_cache.py

```diff
@@ -0,0 +1,297 @@
+
+    meas.register_parameter(DAC.ch1, paramtype=storage_type)
+    meas.register_parameter(DAC.ch2, paramtype=storage_type)
+    meas.register_parameter(DMM.v1, setpoints=(DAC.ch1, DAC.ch2), paramtype=storage_type)
+    array_used = _array_param_used_in_tree(meas)
+    n_rows_written = 0
+    with meas.run(write_in_background=bg_writing) as datasaver:
+        dataset = datasaver.dataset
+        _assert_parameter_data_is_identical(dataset.get_parameter_data(), dataset.cache.data())
+        for v1 in np.linspace(-1, 1, n_points_outer):
+            for v2 in np.linspace(-1, 1, n_points_inner):
+                DAC.ch1.set(v1)
+                DAC.ch2.set(v2)
+                datasaver.add_result((DAC.ch1, v1),
+                                     (DAC.ch2, v2),
+                                     (DMM.v1, DMM.v1.get()))
+                datasaver.flush_data_to_database(block=True)
+                n_rows_written += 1
+                data = dataset.cache.data()
+                if array_used:
+                    shape = (n_rows_written, 1)
+                else:
+                    shape = (n_rows_written,)
+                assert data[DMM.v1.full_name][DMM.v1.full_name].shape == shape
+                assert data[DMM.v1.full_name][DAC.ch1.full_name].shape == shape
+                _assert_parameter_data_is_identical(dataset.get_parameter_data(),
+                                                    data)
+    _assert_parameter_data_is_identical(dataset.get_parameter_data(),
+                                        dataset.cache.data())
+
+
+@pytest.mark.parametrize("bg_writing", [True, False])
+@pytest.mark.parametrize("storage_type", ['numeric', 'array', None])
+@settings(deadline=None, max_examples=10)
+@given(n_points=hst.integers(min_value=1, max_value=21))
+def test_cache_1d_array_in_1d(experiment, DAC, channel_array_instrument, n_points, bg_writing, storage_type):
+    param = channel_array_instrument.A.dummy_array_parameter
+    meas = Measurement()
+    meas.register_parameter(DAC.ch1, paramtype=storage_type)
+    meas.register_parameter(param, setpoints=(DAC.ch1,), paramtype=storage_type)
+    array_used = _array_param_used_in_tree(meas)
+
+    setpoint_name = "_".join((param.instrument.full_name, param.setpoint_names[0]))
+
+    with meas.run(write_in_background=bg_writing) as datasaver:
+        dataset = datasaver.dataset
+        for i, v1 in enumerate(np.linspace(-1, 1, n_points)):
+            datasaver.add_result((DAC.ch1, v1),
+                                 (param, param.get()))
+            datasaver.flush_data_to_database(block=True)
+            data = dataset.cache.data()
+            n_rows_written = i+1
+            if array_used:
+                shape = (n_rows_written, param.shape[0])
+            else:
+                shape = (n_rows_written * param.shape[0],)
+            if storage_type != 'array':
+                # TODO the expansion both for cache and get_parameter_data is buggy iff all types are array
```

Comment:
still todo? what is the bugginess?

Reply:
The expansion is incorrect. This is not a new bug but just revealed by my testing. I will create an issue for it

Reply:
:(